packages/xstate-immer/src/index.ts

```diff
@@ -16,37 +16,51 @@ export interface ImmerAssignAction<TContext, TEvent extends EventObject>
   assignment: ImmerAssigner<TContext, TEvent>;
 }
 
-export function assign<TContext, TEvent extends EventObject = EventObject>(
+function immerAssign<TContext, TEvent extends EventObject = EventObject>(
   assignment: ImmerAssigner<TContext, TEvent>
 ): AssignAction<TContext, TEvent> {
-  // @ts-ignore (possibly infinite TS bug)
   return xstateAssign((context, event) => {
     return produce(context, (draft) => void assignment(draft, event));
   });
 }
 
-interface PatchEventObject extends EventObject {
-  patches: ReturnType<typeof produceWithPatches>;
+export { immerAssign as assign };
+
+export interface ImmerUpdateEventObject extends EventObject {
+  input: any;
 }
 
-export function patchEvent<TContext>(
-  type: string,
-  context: TContext,
-  recipe: (draftContext: TContext) => void
-): PatchEventObject {
-  return {
-    type,
-    patches: produceWithPatches(context, recipe)
-  };
+export interface ImmerUpdateEvent<TType extends string, TInput = any> {
+  type: TType;
+  input: TInput;
 }
 
-export function assignPatch<
-  TContext,
-  TEvent extends PatchEventObject
->(): AssignAction<TContext, TEvent> {
-  return xstateAssign((context, event) => {
-    const [, patches] = event.patches;
+export interface ImmerUpdater<TContext, TEvent extends ImmerUpdateEventObject> {
+  (input: TEvent['input']): TEvent;
+  assign: ImmerAssignAction<TContext, TEvent>;
+  type: TEvent['type'];
+  validate?: (context: TContext, input: TEvent['input']) => boolean;
+}
 
-    return applyPatches(context, patches);
+export function createUpdater<TContext, TEvent extends ImmerUpdateEventObject>(
+  type: TEvent['type'],
+  producer: (ctx: Draft<TContext>, input: TEvent['input']) => void,
+  validate?: (ctx: TContext, input: TEvent['input']) => boolean
```

Comment:
Could u add an example/test for this? Not completely sure what its purpose

Reply:
It's been removed - it's not necessary (you can easily specify guards outside of `createUpdater`)